arrow-libs/core/arrow-autocloseable/src/commonMain/kotlin/arrow/AutoCloseScope.kt

```diff
@@ -0,0 +1,100 @@
+package arrow
+
+import arrow.atomic.Atomic
+import arrow.atomic.update
+
+/**
+ * The [AutoCloseScope] DSL allows for elegantly working with close-ables,
+ * without having to worry about intermediate errors, composing them,
+ * or keeping track of different close-ables and when they need to be closed.
+ *
+ * Similar to C#'s or Scala's Using.
+ * <!--- INCLUDE
+ * import arrow.autoCloseScope
+ * import arrow.install
+ * import java.io.File
+ * import java.io.PrintWriter
+ * import java.util.Scanner
+ * -->
+ * ```kotlin
+ * @ExperimentalStdlibApi
+ * fun main() = autoCloseScope {
+ *   val scanner = install(Scanner(File("testRead.txt")))
+ *   val printer = install(PrintWriter(File("testWrite.txt")))
+ *   for(line in scanner) {
+ *     printer.print(line)
+ *   }
+ * }
+ * ```
+ * <!--- KNIT example-autocloseable-01.kt -->
+ */
+public inline fun <A> autoCloseScope(block: AutoCloseScope.() -> A): A {
+  val scope = DefaultAutoCloseScope()
+  return try {
+    block(scope).also {
+      scope.close(null)
+    }
+  } catch (e: Throwable) {
+    scope.close(e) ?: throw e
```

Comment:
A `nonFatalOrThrow()` seems apt here
```suggestion
    scope.close(e.nonFatalOrThrow()) ?: throw e
```